arangod/Aql/RegisterPlan.cpp

```diff
@@ -320,6 +357,52 @@ auto RegisterPlanT<T>::getTotalNrRegs() -> unsigned int {
   return totalNrRegs;
 }
 
+template <typename T>
+auto RegisterPlanT<T>::calcRegsToKeep(VarSetStack const& varsUsedLaterStack,
+                                      VarSetStack const& varsValidStack,
+                                      std::vector<Variable const*> const& varsSetHere) const
+    -> RegIdSetStack {
+  RegIdSetStack regsToKeepStack;
+  regsToKeepStack.reserve(varsUsedLaterStack.size());
+
+  TRI_ASSERT(varsValidStack.size() == varsUsedLaterStack.size());
+
+  for (auto const [idx, stackEntry] : enumerate(varsValidStack)) {
+    auto& regsToKeep = regsToKeepStack.emplace_back();
+    auto const& varsUsedLater = varsUsedLaterStack[idx];
+
+    for (auto const var : stackEntry) {
+      auto reg = variableToRegisterId(var);
+
+      bool isSetHere = std::find(varsSetHere.begin(), varsSetHere.end(), var) !=
+                       varsSetHere.end();
+      bool isUsedLater = std::find(varsUsedLater.begin(), varsUsedLater.end(), var) !=
+                         varsUsedLater.end();
+      if (!isSetHere && isUsedLater) {
```

Comment:
as a slight optimization, we could save the calculation of `isUsedLater` when `isSetHere` is false, but I am not sure if it is worth it. It _may_ be worth if the function is called often and finding a variable in `varsUsedLater` is expensive.